### DISCUSSION TITLE
Smoothers supported in multigrid

### DISCUSSION THREAD

**Discussion by kiendangtor** [ginkgo-project/ginkgo#1672] [General]
Do you have a document somewhere to show what types of smoothers are currently supported, it would be really good if Schwarz is included. From that works really well with one of my matrix that was generated from an elastic model with multiple interfaces.
Another question is I read sometimes ago that the multiple GPUs on 1 node was planning to be supported so I just want to know if it is currently developed yet? Unexpectedly that distributed was not in the plan but now I saw that part of the code already has distributed parts in it, i.e. Schwarz

**Comment by yhmtsai** [comment]
Yes, preconditioner::Schwarz in ginkgo is already supported in ginkgo multigrid.
the smoother and coarsest solver should accept any kind of LinOp in ginkgo.
you can wrap it in IR with some local solver for example (rough)
`Ir<>::build().with_solver(Schwarz<>::build().with_local_solver(Cg::build()...))...`

The current distribution in ginkgo is 1 gpu per rank. If you have multiple gpus in one physic node, you still need to run it with multiple nodes from mpi such that each node access one gpu. It only applies when you would like to use distributed stuff from ginkgo directly.
we have an example in examples/distributed-solver
Of course, you can still assign different gpu to different executor in one node, run different solver or step, and merge them in end. However, you need to take care of them by yourself.

**Comment by kiendangtor** [comment]
Thank you for your prompt response. I forgot to mention that I would need it without MPI because my main application is not.
 I found Schwarz under experimental::distributed so I built ginkgo with MPI option but then i realized that I need to get distributed exec/matrix for it to work. Is there anyway I can use it without MPI built?

**Reply by yhmtsai** [reply]
without MPI, do you mean using multiple GPUs in a node for a solver?

**Reply by kiendangtor** [reply]
No. I meant can we use it without MPI option? Like sequential build

**Reply by yhmtsai** [reply]
without MPI, you do not need Schwarz.
you can use `Ir<>::build().with_solver(Cg::build()...)` as one of kind of smoother

**Comment by cuihaoran21** [comment]
When I want to solve a large-scale sparse problem with the multigrid Solver, how do I set up its components and parameters? The current setup needs to be iterated too many times      
    ```
using ir = gko::solver::Ir<double>;
    using bj = gko::preconditioner::Jacobi<double, int>;
    using pgm = gko::multigrid::Pgm<double, int>;
    using mg = gko::solver::Multigrid;
 /* Create smoother factory*/
    smoother_gen = gko::share(
        ir::build()
            .with_solver(bj::build().with_max_block_size(32u))
            .with_relaxation_factor(static_cast<double>(0.8))
            .with_criteria(gko::stop::Iteration::build().with_max_iters(10u))
            .on(exec));

    /* Create RestrictProlong factory*/
    mg_level_gen = gko::share(pgm::build().with_deterministic(true).on(exec));

    /* Create CoarsestSolver factory */
    coarsest_solver_gen = gko::share(
    ir::build()
        .with_solver(bj::build().with_max_block_size(32u))
        .with_relaxation_factor(static_cast<double>(0.8))
        .with_criteria(gko::stop::Iteration::build().with_max_iters(1u))
        .on(exec));

    /* Create MultigridSolver factory*/
    multigrid_gen = mg::build()
                        .with_max_levels(10u) //Limit the maximum number of layers for a multigrid
                        .with_min_coarse_rows(4u) //Specifies the minimum number of rows for the coarsest layer
                        .with_pre_smoother(smoother_gen)//Set the pre smoother
                        .with_post_uses_pre(true)//Specify that the post smoother is the same as the pre smoother
                        .with_mg_level(mg_level_gen)//Specifies RestrictProlong
                        .with_coarsest_solver(coarsest_solver_gen)//Sets the coarsest layer solver
                        .with_criteria(iter_stop, tol_stop)//Set stopping criteria including the number of iterations and tolerance
                        .on(exec);
```
